eval.py

```diff
@@ -511,22 +555,87 @@ def evaluate_image( gt: list[WordData],
     gt_occluded = np.asarray([ 'occluded' in el['category'] for el in gt],
                              dtype=bool)
 
+    gt_occluded_visible = np.asarray(
+        ["occluded_english_1" == el["category"] for el in gt], dtype=bool
+    )  # the occluded text is visible text instance in the image
+    gt_occluded_inferable = np.asarray(
+        ["occluded_english_2" == el["category"] for el in gt], dtype=bool
+    )  # the occluded text can be infered from visual cues or common knowledge
+    gt_occluded_indeterminate = np.asarray(
+        ["occluded_english_3" == el["category"] for el in gt], dtype=bool
+    )  # the occluded text can only be infered from the supplementary images
+
     num_tp_occluded = int(np.sum( gt_occluded[matches_gt] ))
     total_gt_occluded = int(np.sum(gt_occluded))
 
+    num_tp_occluded_visible = int(np.sum(gt_occluded_visible[matches_gt]))
+    total_gt_occluded_visible = int(np.sum(gt_occluded_visible))
+
+    num_tp_occluded_inferable = int(np.sum(gt_occluded_inferable[matches_gt]))
+    total_gt_occluded_inferable = int(np.sum(gt_occluded_inferable))
+
+    num_tp_occluded_indeterminate = int(np.sum(gt_occluded_indeterminate[matches_gt]))
+    total_gt_occluded_indeterminate = int(np.sum(gt_occluded_indeterminate))
+
     # Accumulate tightness for matches that count (not ignorable)
     total_tightness = float(np.sum(matches_ious))
 
     results = { 'tp' : int(num_tp),
                 'tp_occluded' : int(num_tp_occluded),
+                'tp_occluded_visible': int(num_tp_occluded_visible),
+                'tp_occluded_inferable': int(num_tp_occluded_inferable),
+                'tp_occluded_indeterminate': int(num_tp_occluded_indeterminate),
                 'total_gt' : int(total_gt),
                 'total_gt_occluded': int(total_gt_occluded),
+                'total_gt_occluded_visible': int(total_gt_occluded_visible),
+                'total_gt_occluded_inferable': int(total_gt_occluded_inferable),
+                'total_gt_occluded_indeterminate': int(total_gt_occluded_indeterminate),
                 'total_pred' :  int(total_pred),
                 'total_tightness' : total_tightness }
 
     stats = get_stats( num_tp, total_gt, total_pred, total_tightness )
-    stats.update( get_occluded_stats( num_tp_occluded, total_gt_occluded,
-                                      stats['precision'] ) )
+    if task == "singledetrec":
```

Comment:
Same note here about comments. But now that these very similar blocks of code are much larger, where they used to each be a "one liner", I think it could benefit from factorization in to a common body so they retain that "one liner" essence.  This instance modifies `stats` with the four(?) raw counts, and the other instance above modifies `final_stats`. The parameters differ only slightly, but the structure is otherwise the same, which is a good reason for a new higher-level procedure to aid readability, maintainability, and correctness.